### DISCUSSION TITLE
VIPS runs out of memory (~20Gb) while calling dzsave() on very large image

### DISCUSSION THREAD

**Discussion by harryrobbins** [libvips/libvips#2373] [Q&A]
Firstly, thanks for all your amazing hard work on this library - I'd be completely stuck without it !

# Goal

I'm trying to stitch together 50,000 one megapixel .jpegs (1000px x 1000px) into a large image, and then generate Deep Zoom tiles that I can view in Open Seadragon. 

I'd like to use Python (ideally) or the command line if that's not possible.

Apologies if this info should be in the `pyvips` repository but I thought that, as I also encounter memory issues on the command line, this might be the right place. 

# Problem 

When I try to stitch together more that 20,000 images and turn them into tiles I run out of memory and the process crashes.

I realise the 50 gigapixel target is a very large image and possibly larger than this library was intended to support but, from reading your comments on other issues, I got the impressions that the dzsave() method should not be overly memory-hungry so perhaps I'm doing something wrong.

## Code

```python3
def process_size():
    print(f"process size: {process.memory_info().rss / (1024 * 1024):.2f} MB")

print("Starting VIPS processes\n")
print(f"pyvips version: {pyvips.__version__}")
print(f"libvips version: {os.system('vips --version')}")
process_size()


for image_path in image_paths:
    image = pyvips.Image.new_from_file(str(image_path), access='sequential')
    if (image.height >= 1000) and (image.width >= 1000):
        images.append(image.crop(0,0, 1000,1000 ))
		
process_size()

print(f"\nStitching massive image from {len(images)} images")
massive_image = pyvips.Image.arrayjoin(images, across=250)
process_size()

dz_dir = Path("./data/dz/")
dz_dir.mkdir(parents=True, exist_ok=True)


print(f"\nSaving tiles to {dz_dir}")
massive_image.dzsave("massive_stitch", dirname=str(dz_dir), depth='onetile')
process_size()
```

## Output
```shell
Starting VIPS processes

pyvips version: 2.1.15
libvips version: vips-8.11.2
process size: 47.14 MB


process size: 123.20 MB

Stitching massive image from 30000 images
process size: 245.41 MB

Saving tiles to data/dz
Segmentation fault
```

On one run I got a slightly different error:
```
Stitching massive image from 50000 images
process size: 376.28 MB

Saving tiles to data/dz
Traceback (most recent call last):
  File "join_and_tile.py", line 124, in <module>
    massive_image.dzsave("massive_stitch", dirname=str(dz_dir), depth='onetile')
  File "/home/harry/miniconda3/lib/python3.8/site-packages/pyvips/vimage.py", line 921, in call_function
    return pyvips.Operation.call(name, self, *args, **kwargs)
  File "/home/harry/miniconda3/lib/python3.8/site-packages/pyvips/voperation.py", line 293, in call
    raise Error('unable to call {0}'.format(operation_name))
pyvips.error.Error: unable to call dzsave
  glib: Error creating thread: Resource temporarily unavailable

glib: Error creating thread: Resource temporarily unavailable

glib: Error creating thread: Resource temporarily unavailable

glib: Error creating thread: Resource temporarily unavailable

wbuffer_write: write failed

```

During the "Saving tiles to data/dz" stage it quickly uses up all the free system memory, runs for about 30 minutes, and then crashes.

With 10,000 or 20,000 images it works fine and generates the images in about 15 mins/30 mins respectively.

## Setup

I'm running the code on a (fairly old) workstation with:

* 24 Gb RAM
* 12 Xeon cores (X5670  @ 2.93GHz)
* Ubuntu 20.04 virtualised in Windows (via WSL 2)
* `pyvips 2.1.15`
* `libvips vips-8.11.2`
* `python 3.8`

I also have various versions of libjpeg installed - maybe this is an issue?
```shell
$ apt list --installed |grep jpeg
libjpeg-turbo8/focal-updates,focal-security,now 2.0.3-0ubuntu1.20.04.1 amd64 [installed,automatic]
libjpeg8/focal,now 8c-2ubuntu8 amd64 [installed,automatic]
libturbojpeg/focal-updates,focal-security,now 2.0.3-0ubuntu1.20.04.1 amd64 [installed]
```

## Alternative approach using only command line

I tried the following in a directory containing 2,682 jpegs (roughly, but not exactly, 1000px x 1000px)

```
$ ls -lR . |  wc -l
2682
$ vips arrayjoin "$(echo *.jpg)" ../mypyramid.dz --across 100  --vips-progress --vips-leak 
```

Which produce a lot of this very quickly:
```
vips temp-24093: 1 x 1 pixels, 24 threads, 1 x 1 tiles, 512 lines in buffer
vips temp-24093: done in 0.000589s
```

and then

```
vips temp-24094: 108000 x 53136 pixels, 24 threads, 108000 x 16 tiles, 512 lines in buffer
vips temp-24094: 49% complete
```

while it was running `top` output the following:

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
15360 harry     20   0   41.4g  18.2g   1084 S  1121  96.9  51:45.36 vips
```

And then at 74% it output
```
vips temp-24094: 108000 x 53136 pixels, 24 threads, 108000 x 16 tiles, 512 lines in buffer
Killedemp-24094: 74% complete
```

Any ideas what I might do to avoid these issues (ideally via `pyvips`)?

I've ordered some more RAM (just in case) but I thought that might not be an option for all users and there might be a more elegant solution.

**Comment by jcupitt** [comment]
Hello @harryrobbins,

I think the high memory use you are seeing is from file read rather than from dzsave. The "how libvips opens files" chapter has the details:

https://libvips.github.io/libvips/API/current/How-it-opens-files.md.html

For jpeg in sequential mode, it'll stream from each file, but it'll buffer a fair number of scanlines. These buffers build up in memory when you have a large number of inputs.

I tried your example:

```
john@banana ~/pics/x $ vips crop ../k2.jpg x.jpg 0 0 1000 1000
john@banana ~/pics/x $ for i in {1..3000}; do cp x.jpg $i.jpg; done
john@banana ~/pics/x $ vips arrayjoin "$(echo *.jpg)" x.dz[overlap=0] --across 100 --vips-progress --vips-leak
vips temp-6010: 1 x 1 pixels, 32 threads, 1 x 1 tiles, 640 lines in buffer
vips temp-6010: done in 0.000505s          
vips temp-6011: 100000 x 31000 pixels, 32 threads, 100000 x 16 tiles, 640 lines in buffer
vips temp-6011: done in 65.3s          
memory: high-water mark 30.67 GB
```

(this PC has 128gb)

So it works, but it does need a lot of memory. If you use a format like .v instead, which does not need buffering, it's better:

```
john@banana ~/pics/x $ vips crop ../k2.jpg x.v 0 0 1000 1000
john@banana ~/pics/x $ for i in {1..3000}; do cp x.v $i.v; done
john@banana ~/pics/x $ vips arrayjoin "$(echo *.v)" x.dz[overlap=0] --across 100 --vips-progress --vips-leak
vips temp-6010: 1 x 1 pixels, 32 threads, 1 x 1 tiles, 640 lines in buffer
vips temp-6010: done in 0.00046s          
vips temp-6011: 100000 x 31000 pixels, 32 threads, 100000 x 1 tiles, 640 lines in buffer
vips temp-6011: done in 57.2s          
memory: high-water mark 535.83 MB
```

But now you need a huge amount of disk instead, since .v is uncompressed.

There's probably something we can do to reduce peak memory in this case. Maybe for 8.12.

**Reply by harryrobbins** [reply]
Amazing - thanks so much for the detailed answer and quick response. 

I got it to run eventually on a server with lots of memory so will probably continue with that approach for now. Glad to know I'm  not missing an obvious trick.
